src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,151 @@
+        );
+      },
+      tooManySites: (count) => {
+        return i18n.sprintf(i18n.ngettext(
+          'Access your data on %(count)s other site',
+          'Access your data on %(count)s other sites',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+    };
+    this.hostPermissions = [];
+  }
+
+  permissionStrings: Object;
+  hostPermissions: Array<React.Element<typeof Permission>>;
+
+  // Add individual Permission components for individual site/domain permissions.
+  formatItems(items: Array<string>, messageType: string) {
+    for (const item of items) {
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings[messageType](item)}
+          key={item}
+        />
+      );
+    }
+  }
+
+  // Adds Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  addHostPermissions(
+    list: Array<string>,
+    singleMessageType:string,
+    moreMessageType:string
+  ) {
+    if (list.length < 5) {
+      this.formatItems(list, singleMessageType);
+    } else {
+      this.formatItems(list.slice(0, 3), singleMessageType);
+
+      const remaining = list.length - 3;
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings[moreMessageType](remaining)}
+          key={moreMessageType}
+        />
+      );
+    }
+  }
+
+  render() {
+    const { permissions } = this.props;
+
+    // Group permissions into "site" and "domain" permissions. If any
+    // "all urls" permissions are found, break as we'll only need one
+    // host permissions message.
+    let allUrls = false;
+    const wildcards = [];
+    const sites = [];
+    for (const permission of permissions) {
+      if (permission === '<all_urls>') {
+        allUrls = true;
+        break;
+      }
+      if (permission.startsWith('moz-extension:')) {
+        continue;
+      }
+      const match = /^[a-z*]+:\/\/([^/]+)\//.exec(permission);
+      if (!match) {
+        log.debug(`Host permission string "${permission}" appears to be invalid.`);
+        continue;
+      }
+      if (match[1] === '*') {
+        allUrls = true;
+      } else if (match[1].startsWith('*.')) {
+        wildcards.push(match[1].slice(2));
+      } else {
+        sites.push(match[1]);
+      }
+    }
+
+    // Format the host permissions.  If we have a wildcard for all urls,
+    // a single string will suffice.  Otherwise, show domain wildcards
+    // first, then individual host permissions.
+    if (allUrls) {
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings.allUrls}
+          key="allUrls"
+        />
+      );
+    } else {
+      this.addHostPermissions(wildcards, 'wildcard', 'tooManyWildcards');
```

Comment:
While flow (in theory when used everywhere in our app) prevents us from sending the wrong magic constant to arguments by strictly defining accepted argument values, I still find this a hard API to read and it's still a magic constant. Using a constant variable instead of a string directly makes ambiguities, however rare they may be, less likely to crop up. I also think they're just more idiomatic to me as a programmer, but that might be preference rather than a real best practice.

At any rate, seeing a few string arguments like this makes it hard for me to reason about what this API does. An object as an argument might illuminate things more, eg:

```es6
addHostPermission({
  permissions: wildcards,
  messageType: 'wildcard',
});
```

I'll point out here too that I don't think we need separate "single" and "tooMany" message types. The only time you use the "tooMany" one is a code path of having too many items so you can handle that logic in that function without exposing it elsewhere, if that makes sense.